mono/mini/mini-amd64.c

```diff
@@ -392,6 +392,20 @@ add_valuetype_win64 (MonoMethodSignature *sig, ArgInfo *ainfo, MonoType *type,
 
 	klass = mono_class_from_mono_type (type);
 	size = mini_type_stack_size_full (&klass->byval_arg, NULL, sig->pinvoke);
+
+	ainfo->pass_empty_struct = FALSE;
+#ifndef __GNUC__
+	/*
+	* Standard C and C++ doesn't allow empty structs, empty structs will always have a size of 1 byte.
+	* GCC have an extension to allow empty structs, https://gcc.gnu.org/onlinedocs/gcc/Empty-Structures.html.
+	* This cause a little dilemma since runtime build using none GCC compiler will not be compatible with
+	* GCC build C libraries and the other way around. On platforms where empty structs has size of 1 byte
+	* it must be represented in call and cannot be dropped.
+	*/
+	if (0 == size && MONO_TYPE_ISSTRUCT (type) && sig->pinvoke)
+		ainfo->pass_empty_struct = TRUE;
+#endif
```

Comment:
I piggy back on the logic currently based on storage in order to allocate registers or stack slots for the empty struct elsewhere. Using a separate parameter will keep that logic intact and I can extend on it just in the location where we need to handle things differently. The pass_empty_struct also keeps logic around status of pinvoke since it will only be needed in that scenario. Instead of putting similar conditions in several other places, I simplified it into this bool making it esier for other code to validate the condition regardless what actually triggered it to be true. If it’s OK I would like to keep it as is.